pkg/parquetquery/iters.go

```diff
@@ -1368,68 +1366,59 @@ func (j *JoinIterator) String() string {
 }
 
 func (j *JoinIterator) Next() (*IteratorResult, error) {
-	// Here is the algorithm for joins:  On each pass of the iterators
-	// we remember which ones are pointing at the earliest rows. If all
-	// are the lowest (and therefore pointing at the same thing) then
-	// there is a successful join and return the result.
-	// Else we progress the iterators and try again.
-	// There is an optimization here in that we can seek to the highest
-	// row seen. It's impossible to have joins before that row.
+outer:
 	for {
-		lowestRowNumber := MaxRowNumber()
-		highestRowNumber := EmptyRowNumber()
-		j.lowestIters = j.lowestIters[:0]
+		// This loop is doing two things:
+		// On first-pass peek each iter and ensure it has at least one
+		// result.  If any iter has no results we can exit early
+		// without processing the remaining data in the others.
+		// On subsequent passes the first iter is never nil except
+		// when everything is fully exhausted. We check once more
+		// and then exit.
+		if j.peeks[0] == nil {
+			for i := range j.iters {
+				res, err := j.peek(i)
+				if err != nil {
+					return nil, err
+				}
+				if res == nil {
+					return nil, nil
+				}
+			}
+		}
 
-		for iterNum := range j.iters {
-			res, err := j.peek(iterNum)
+		// The first iter is pointing at the next candidate row. Proceed through iters 2 to N looking
+		// for matches.
+		for iterNum := 1; iterNum < len(j.iters); iterNum++ {
+			err := j.seek(iterNum, j.peeks[0].RowNumber, j.definitionLevel)
 			if err != nil {
-				return nil, fmt.Errorf("join iterator peek failed: %w", err)
+				return nil, err
 			}
 
-			if res == nil {
-				// Iterator exhausted, no more joins possible
+			if j.peeks[iterNum] == nil {
 				return nil, nil
 			}
 
-			c := CompareRowNumbers(j.definitionLevel, res.RowNumber, lowestRowNumber)
-			switch c {
-			case -1:
-				// New lowest, reset
-				j.lowestIters = j.lowestIters[:0]
-				lowestRowNumber = res.RowNumber
-				fallthrough
-
-			case 0:
-				// Same, append
-				j.lowestIters = append(j.lowestIters, iterNum)
-			}
-
-			if CompareRowNumbers(j.definitionLevel, res.RowNumber, highestRowNumber) == 1 {
-				// New high water mark
-				highestRowNumber = res.RowNumber
+			if CompareRowNumbers(j.definitionLevel, j.peeks[iterNum].RowNumber, j.peeks[0].RowNumber) == 1 {
+				// This iterator has a higher row number than all previous iterators.  That means it might have
+				// a higher filtering power, swap it to the top and restart the loop.
+				j.iters[0], j.iters[iterNum] = j.iters[iterNum], j.iters[0]
```

Comment:
this swap has me thinking about other options. for instance, we could move the new primary iterator to the front. presumably this would mean that by the end of the file the iters would be sorted by effectiveness.

difficult to prove this is better than the swap here.

Reply:
Sorry. That wasn't very descriptive. I meant:

1) Identify new primary iterator at index N
2) Save off the new primary iterator
3) Shift index 0 through N-1 up one slot
4) Copy the new primary iterator into 0

So an iterator that was previously the primary iterator will remain toward the front.

Like i said in the other comment. There's infinite depth here and it would easy to overtune for one query or block. I'm more than fine with the PR as is.